lte/gateway/c/session_manager/ShardTracker.cpp

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string>
+#include <vector>
+
+#include "ShardTracker.h"
+
+namespace magma {
+
+int ShardTracker::add_ue() {
+  int shard_id;
+  // If there are no shards, add an entry for the shard
+  // representing a single UE at the new shard
+  if (ue_count_per_shard_.size() == 0) {
+    ue_count_per_shard_.push_back(1);
+    return 0;
+  } else {
+    // Iterate through all existing shards, if all are full
+    // create a new shard with quantity 1, otherwise increment
+    // the quantity of UEs in the latest shard
+    for (auto i = 0; i < ue_count_per_shard_.size(); i++) {
+      if (ue_count_per_shard_[i] < max_shard_size_) {
+        ue_count_per_shard_[i]++;
+        return i;
+      }
+    }
+    shard_id = ue_count_per_shard_.size();
+    ue_count_per_shard_.push_back(1);
```

Comment:
If we assign a UE by adding a new shard, can't we just return the new shard id here? Do we need to keep going with the loop?

Reply:
If we have 
* max_shard_size_ = 5
* shard ID 0 with 5UEs
* shard ID 1 with 1UE

This would add a new shard ID 2 with 1UE when `add_ue` is called, is that right?

Reply:
Correct. But this handles the case where we need to add a new shard id. Meaning all the shard ids elsewhere are already filled. So like if shard ID 0 has 5 UEs and shard ID one has 5 UEs, then we create new shard id 2, and add 1 UE.